src/common/pagination.py

```diff
@@ -12,102 +12,162 @@
-    materials_manager = aws_encryption_sdk.CachingCryptoMaterialsManager(
-        master_key_provider=kms_key_provider,
-        cache=cache,
-        max_age=cache_max_age.total_seconds(),
-    )
-
-    return EncryptionClient(
-        client=client,
-        key_provider=kms_key_provider,
-        cache=cache,
-        materials_manager=materials_manager,
-    )
-
-
-def _decode_plaintext_pagination_token(pagination_token: str) -> dict:
+        return token
+
+    def _validate(
+        self,
+        *,
+        expiration: datetime,
+        context: dict,
+        disable_strict_validation: bool,
+        now: datetime,
+    ):
+        if expiration < now:
+            raise ValueError(f"Token expired")
+
+        for key in self.context.keys():
+            if key not in context:
+                if disable_strict_validation:
+                    continue
+                raise KeyError(f"Missing key '{key}'")
+            if context[key] != self.context[key]:
+                raise ValueError(f"Mismatch in key '{key}'")
+
+
+def pagination_token_context_from_event(
+    event: dict, *, duration: timedelta, now: Optional[datetime] = None
+) -> PaginationTokenContext:
+    """Create a context from the current event and a token validity duration."""
+    context = {
+        # Ensure that a token issued for one user can't be used by another user.
+        "caller_identity": get_caller_identity(event),
+        # Ensure that tokens issued for one API can't be used in another API
+        "api_id": get_api_id(event),
+        # Ensure that a token issued for one resource type / path can't be used for
+        # another resource type.
+        "path": get_path(event),
+    }
+
+    if now is None:
+        now = datetime.now(timezone.utc)
+    expiration = now + duration
+
+    return PaginationTokenContext(context=context, expiration=expiration)
+
+
+def _decode_plaintext_pagination_token(
+    pagination_token: str, context: PaginationTokenContext
+) -> dict:
     pagination_token_bytes = base64.urlsafe_b64decode(pagination_token)
-    parsed_pagination_token = json.loads(pagination_token_bytes)
-    return parsed_pagination_token
+    parsed_pagination_token_data = json.loads(pagination_token_bytes)
+    pagination_token = context.load(parsed_pagination_token_data)
+    return pagination_token
 
 
-def _encode_plaintext_pagination_token(last_evaluated_key: dict) -> str:
-    serialized_last_evaluated_key = json.dumps(last_evaluated_key)
-    pagination_token_bytes = base64.urlsafe_b64encode(
-        serialized_last_evaluated_key.encode("ascii")
-    )
+def _encode_plaintext_pagination_token(
+    pagination_token_data: Any, context: PaginationTokenContext
+) -> str:
+    packaged_data = context.package(pagination_token_data)
+    serialized_data = json.dumps(packaged_data, ensure_ascii=True)
+    pagination_token_bytes = base64.urlsafe_b64encode(serialized_data.encode("ascii"))
```

Comment:
The `ascii` bits here surprised me, as my understanding is that JSON is canonically UTF-8. Today I learned that `json.dumps(..., ensure_ascii=True)` is a thing ([and the default behaviour](https://docs.python.org/3/library/json.html#character-encodings)). This feels like something you've been burned by in the past 🙂 Tucking this away for future pondering.

Reply:
I guess we could do it either way.